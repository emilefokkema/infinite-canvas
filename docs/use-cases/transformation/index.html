<html>
<head>
    <title>With different width and height</title>
</head>
<body>
<div class="canvas-block">
    <canvas id="canvas"></canvas>
    <canvas id="left-gutter"></canvas>
    <canvas id="top-gutter"></canvas>
</div>
</body>
</html>
