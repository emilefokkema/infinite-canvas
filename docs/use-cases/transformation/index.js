import './style.css'
import InfiniteCanvas from 'infinite-canvas'

function getMarkings(start, length){
    function* getMarkings(stepSize){
        var firstStep = stepSize * Math.ceil(start / stepSize);
        for(let loc = firstStep; loc <= start + length; loc += stepSize){
            yield loc;
        }
    }
    var stepSize = Math.pow(10, Math.floor(Math.log(length) / Math.log(10) - 0.1));

    function* getMainMarkings(){
        yield* getMarkings(stepSize);
    }
    function* getSubMarkings(){
        yield* getMarkings(stepSize / 10);
    }
    return {getMainMarkings, getSubMarkings};
}
var leftGutter = (function(){
    var width = 40;
    var topOffset = 20;
    var canvas = document.getElementById("left-gutter");
    canvas.width = width;
    var ctx = canvas.getContext("2d");
    var gradient = ctx.createLinearGradient(0, 0, 40, 0);
    gradient.addColorStop(0, '#eee');
    gradient.addColorStop(1, '#bbb');
    function drawMarkedTick(location, inverseTransformation){
        var transformedLocation = location * inverseTransformation.d + inverseTransformation.f - topOffset;
        ctx.fillStyle = '#000';
        ctx.fillRect(25, transformedLocation, 15, 1);
        ctx.fillText(location, 10, transformedLocation);
    }
    function drawSubTick(location, inverseTransformation){
        var transformedLocation = location * inverseTransformation.d + inverseTransformation.f - topOffset;
        ctx.fillStyle = '#000';
        ctx.fillRect(35, transformedLocation, 5, 1);
    }
    function drawWithHeightAndTransformation(height, transformation, inverseTransformation){
        height = height - topOffset;
        canvas.height = height;
        ctx.fillStyle = gradient;
        ctx.fillRect(0, 0, width, height);
        var topCoordinate = topOffset * transformation.d + transformation.f;
        var distance = height * transformation.d;
        var markings = getMarkings(topCoordinate, distance);
        for(let loc of markings.getMainMarkings()){
            drawMarkedTick(loc, inverseTransformation);
        }
        for(let loc of markings.getSubMarkings()){
            drawSubTick(loc, inverseTransformation);
        }
    }
    return {drawWithHeightAndTransformation};
})()
var topGutter = (function(){
    var height = 20;
    var canvas = document.getElementById("top-gutter");
    canvas.height = height;
    var ctx = canvas.getContext("2d");
    var gradient = ctx.createLinearGradient(0, 0, 0, 20);
    gradient.addColorStop(0, '#eee');
    gradient.addColorStop(1, '#bbb');
    function drawMarkedTick(location, inverseTransformation){
        var transformedLocation = location * inverseTransformation.a + inverseTransformation.e;
        ctx.fillStyle = '#000';
        ctx.fillRect(transformedLocation, 10, 1, 10);
        ctx.fillText(location, transformedLocation - 10, 8)
    }
    function drawSubTick(location, inverseTransformation){
        var transformedLocation = location * inverseTransformation.a + inverseTransformation.e;
        ctx.fillStyle = '#000';
        ctx.fillRect(transformedLocation, 15, 1, 5);
    }
    function drawWithWidthAndTransformation(width, transformation, inverseTransformation){
        canvas.width = width;
        ctx.fillStyle = gradient;
        ctx.fillRect(0, 0, width, height);
        var leftCoordinate = transformation.e;
        var distance = width * transformation.a;
        var markings = getMarkings(leftCoordinate, distance);
        for(let loc of markings.getMainMarkings()){
            drawMarkedTick(loc, inverseTransformation);
        }
        for(let loc of markings.getSubMarkings()){
            drawSubTick(loc, inverseTransformation);
        }
    }
    return {drawWithWidthAndTransformation};
})();
var canvasEl = document.getElementById("canvas");
canvasEl.width = 1200;
canvasEl.height = 1200;
var infCanvas = new InfiniteCanvas(canvasEl, {greedyGestureHandling: true, rotationEnabled: false, units: InfiniteCanvas.CSS_UNITS});
var ctx = infCanvas.getContext();
infCanvas.addEventListener("draw", (ev) => {
    var rect = canvasEl.getBoundingClientRect();
    topGutter.drawWithWidthAndTransformation(rect.width, ev.transformation, ev.inverseTransformation);
    leftGutter.drawWithHeightAndTransformation(rect.height, ev.transformation, ev.inverseTransformation);
});
function drawDot(x, y){
    ctx.fillStyle = '#000';
    ctx.beginPath();
    ctx.arc(x, y, 1, 0, 2 * Math.PI);
    ctx.fill();
    ctx.fillText(`(${x}, ${y})`, x - 5, y - 5);
}
drawDot(0, 0)
drawDot(200, 0)
drawDot(0, 200)
ctx.fillStyle = '#aaa';
ctx.rotate(Math.PI / 8)
ctx.fillRect(50, 50, Infinity, 100);
ctx.strokeRect(50, 50, Infinity, 100);
ctx.fillStyle = '#faa';
ctx.translate(100, 50);
ctx.rotate(Math.PI / 8);
ctx.fillRect(-50, -100, Infinity, 100);
ctx.strokeRect(-50, -100, Infinity, 100);