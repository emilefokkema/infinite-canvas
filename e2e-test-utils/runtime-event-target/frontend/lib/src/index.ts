import type { EventTargetFactoryInitializer } from '../../api/event-target-factory-initializer'
import { RuntimeEventTargetImpl } from './runtime-event-target-impl'
import { ExposedFunctionConnection } from './exposed-function-connection'

const initializeEventTargetFactory: EventTargetFactoryInitializer = () => {
    return { 
        createEventTarget(eventTargetLike, connectionData) {
            const connection = new ExposedFunctionConnection(connectionData);
            return new RuntimeEventTargetImpl(connection, eventTargetLike);
        }
    };

}

export default initializeEventTargetFactory;