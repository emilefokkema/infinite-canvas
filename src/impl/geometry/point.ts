export class Point{
	constructor(public readonly x: number, public readonly y: number){}
	public mod(): number{
		return Math.sqrt(this.modSq());
	}
	public modSq(): number{
		return this.x * this.x + this.y * this.y;
	}
	public minus(other: Point): Point{
		return new Point(this.x - other.x, this.y - other.y);
	}
	public plus(other: Point): Point{
		return new Point(this.x + other.x, this.y + other.y);
	}
	public dot(other: Point): number{
		return this.x * other.x + this.y * other.y;
	}
	public cross(other: Point): number{
		return this.x * other.y - this.y * other.x;
	}
	public equals(other: Point): boolean{
		return this.x === other.x && this.y === other.y;
	}
	public getPerpendicular(): Point{
		return new Point(-this.y, this.x);
	}
	public scale(r: number): Point{
		return new Point(r * this.x, r * this.y);
	}
	public projectOn(other: Point): Point{
		return other.scale(this.dot(other) / other.modSq());
	}
	public matrix(a: number, b: number, c: number, d: number): Point{
		return new Point(a * this.x + b * this.y, c * this.x + d * this.y);
	}
	public inSameDirectionAs(other: Point): boolean{
		return this.cross(other) === 0 && this.dot(other) >= 0;
	}
	public isInOppositeDirectionAs(other: Point): boolean{
		return this.cross(other) === 0 && this.dot(other) < 0;
	}
	public isOnSameSideOfOriginAs(other1: Point, other2: Point){
		return this.isInSmallerAngleBetweenPoints(other1, other2) || 
			   other1.isInSmallerAngleBetweenPoints(this, other2) ||
			   other2.isInSmallerAngleBetweenPoints(this, other1);
	}
	public isInSmallerAngleBetweenPoints(point1: Point, point2: Point): boolean{
		const cross: number = point1.cross(point2);
		if(cross > 0){
			return point1.cross(this) >= 0 && this.cross(point2) >= 0;
		}else if(cross < 0){
			return point1.cross(this) <= 0 && this.cross(point2) <= 0;
		}else if(point1.dot(point2) > 0){
			return this.cross(point1) === 0 && this.dot(point1) > 0;
		}else{
			return true;
		}
	}
	public static origin: Point = new Point(0, 0);
}