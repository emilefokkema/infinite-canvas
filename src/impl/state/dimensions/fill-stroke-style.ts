import { InfiniteCanvasFillStrokeStyle } from "../../styles/infinite-canvas-fill-stroke-style";
import { Instruction } from "../../instructions/instruction";
import { TypedStateInstanceDimension } from "./typed-state-instance-dimension";
import { InfiniteCanvasStateInstance } from "../infinite-canvas-state-instance";
import { InfiniteCanvasPattern } from "../../styles/infinite-canvas-pattern";

class FillStrokeStyle<K extends "fillStyle" | "strokeStyle"> implements TypedStateInstanceDimension<string | CanvasGradient | CanvasPattern>{
    constructor(private readonly propName: K){}
    public changeInstanceValue(instance: InfiniteCanvasStateInstance, newValue: string | CanvasGradient | CanvasPattern): InfiniteCanvasStateInstance{
        return instance.changeProperty(this.propName, newValue);
    }
    public isEqualForInstances(one: InfiniteCanvasStateInstance, other: InfiniteCanvasStateInstance): boolean{
        return one[this.propName] === other[this.propName];
    }
    public getInstructionToChange(fromInstance: InfiniteCanvasStateInstance, toInstance: InfiniteCanvasStateInstance): Instruction{
        const newValue: string | CanvasGradient | CanvasPattern = toInstance[this.propName];
        if(this.isEqualForInstances(fromInstance, toInstance)){
            if(!(newValue instanceof InfiniteCanvasFillStrokeStyle) || fromInstance.fillAndStrokeStylesTransformed === toInstance.fillAndStrokeStylesTransformed){
                return () => {};
            }
            return toInstance.fillAndStrokeStylesTransformed ? newValue.getInstructionToSetTransformed(this.propName) : newValue.getInstructionToSetUntransformed(this.propName);
        }
        if(newValue instanceof InfiniteCanvasFillStrokeStyle){
            return toInstance.fillAndStrokeStylesTransformed ? newValue.getInstructionToSetTransformed(this.propName) : newValue.getInstructionToSetUntransformed(this.propName);
        }
        

        return (context: CanvasRenderingContext2D) => {context[this.propName] = toInstance[this.propName];};
    }
    public valueIsTransformableForInstance(instance: InfiniteCanvasStateInstance): boolean{
        return !(instance[this.propName] instanceof InfiniteCanvasPattern);
    }
}
export const fillStyle = new FillStrokeStyle("fillStyle");
export const strokeStyle = new FillStrokeStyle("strokeStyle");
