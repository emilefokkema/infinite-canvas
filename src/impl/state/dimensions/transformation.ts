import { Transformation } from "../../transformation";
import { InfiniteCanvasStateInstanceDimension } from "./infinite-canvas-state-instance-dimension";
import { Instruction, noopInstruction } from "../../instructions/instruction";
import { TypedStateInstanceDimension } from "./typed-state-instance-dimension";

class TransformationDimension extends InfiniteCanvasStateInstanceDimension<"transformation">{
    protected valuesAreEqual(oldValue: Transformation, newValue: Transformation): boolean{
        return oldValue.equals(newValue);
    }
    protected changeToNewValue(newValue: Transformation): Instruction{
        return (ctx, rectangle) => {
            const {a, b, c, d, e, f} = rectangle.getTransformationForInstruction(newValue);
            ctx.setTransform(a, b, c, d, e, f)
        }
    }
}
export const transformation: TypedStateInstanceDimension<Transformation> = new TransformationDimension("transformation", noopInstruction);