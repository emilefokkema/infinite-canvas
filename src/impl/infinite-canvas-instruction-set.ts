import { InfiniteCanvasState } from "./state/infinite-canvas-state";
import { Instruction } from "./instructions/instruction";
import { InfiniteCanvasStateInstance } from "./state/infinite-canvas-state-instance";
import { PreviousInstructions } from "./instructions/previous-instructions";
import { CurrentPath } from "./interfaces/current-path";
import { InstructionsWithPath } from "./instructions/instructions-with-path";
import { PathInstruction } from "./interfaces/path-instruction";
import { TransformationKind } from "./transformation-kind";
import { Area } from "./areas/area";
import { empty } from "./areas/empty";
import { Position } from "./geometry/position"
import { ExecutableInstructionWithState } from "./instructions/executable-instruction-with-state";
import { InstructionsWithPositiveDrawnArea } from "./instructions/instructions-with-positive-drawn-area";
import { ExecutableStateChangingInstructionSet } from "./interfaces/executable-state-changing-instruction-set";
import { DrawingInstruction } from './drawing-instruction'
import { CanvasRectangle } from "./rectangle/canvas-rectangle";
import { roundRect } from "./rect/round-rect";
import { getRectStrategy} from "./rect/get-rect-strategy";
import { Rect } from './rect/rect'

export class InfiniteCanvasInstructionSet{
    private currentInstructionsWithPath: CurrentPath;
    private previousInstructionsWithPath: PreviousInstructions;
    public state: InfiniteCanvasState;
    constructor(private readonly onChange: () => void){
        this.previousInstructionsWithPath = PreviousInstructions.create();
        this.state = this.previousInstructionsWithPath.state;
    }
    public beginPath(): void{
        const newCurrentPath = InstructionsWithPath.create(this.state);
        newCurrentPath.setInitialStateWithClippedPaths(this.previousInstructionsWithPath.state);
        this.currentInstructionsWithPath = newCurrentPath;
    }
    public changeState(change: (state: InfiniteCanvasStateInstance) => InfiniteCanvasStateInstance): void{
        this.state = this.state.withCurrentState(change(this.state.current));
    }
    public saveState(): void{
        this.state = this.state.saved();
    }
    public restoreState(): void{
        this.state = this.state.restored();
    }
    public resetState(): void{
        this.previousInstructionsWithPath = PreviousInstructions.create();
        this.state = this.previousInstructionsWithPath.state;
        this.currentInstructionsWithPath = undefined;
        this.onChange();
    }
    public allSubpathsAreClosable(): boolean{
        return !this.currentInstructionsWithPath || this.currentInstructionsWithPath.allSubpathsAreClosable();
    }
    public currentPathSurroundsFinitePoint(): boolean{
        return this.currentInstructionsWithPath && this.currentInstructionsWithPath.surroundsFinitePoint();
    }
    public currentSubpathIsClosable(): boolean{
        return !this.currentInstructionsWithPath || this.currentInstructionsWithPath.currentSubpathIsClosable();
    }
    public fillPath(instruction: Instruction): void{
        if(!this.currentInstructionsWithPath){
            return;
        }
        const drawingInstruction = DrawingInstruction.forFillingPath(instruction, this.state, () => this.currentInstructionsWithPath)
        this.state = drawingInstruction.state;
        this.incorporateDrawingInstruction(drawingInstruction)
    }
    public strokePath(): void{
        if(!this.currentInstructionsWithPath){
            return;
        }
        const drawingInstruction = DrawingInstruction.forStrokingPath((context) => {context.stroke();}, this.state, () => this.currentInstructionsWithPath)
        this.state = drawingInstruction.state;
        this.incorporateDrawingInstruction(drawingInstruction)
    }
    public fillRect(rect: Rect, instruction: Instruction): void{
        const drawingInstruction = rect.fill(this.state, instruction)
        if(!drawingInstruction){
            return
        }
        this.incorporateDrawingInstruction(drawingInstruction);
    }

    public strokeRect(rect: Rect): void{
        const drawingInstruction = rect.stroke(this.state, (context) => {context.stroke();})
        if(!drawingInstruction){
            return
        }
        this.incorporateDrawingInstruction(drawingInstruction);
    }

    public addDrawing(
        instruction: Instruction,
        area: Area,
        transformationKind: TransformationKind,
        takeClippingRegionIntoAccount: boolean,
        tempStateFn: (state: InfiniteCanvasStateInstance) => InfiniteCanvasStateInstance): void{
            const state = this.state.currentlyTransformed(false);
            if(transformationKind === TransformationKind.Relative){
                area = area.transform(this.state.current.transformation);
            }
            let tempState: InfiniteCanvasState;
            if(tempStateFn){
                tempState = state.withCurrentState(tempStateFn(state.current));
            }
            this.incorporateDrawingInstruction(new DrawingInstruction(
                instruction,
                area,
                (instruction) => {
                    return ExecutableInstructionWithState.create(state, instruction)
                },
                takeClippingRegionIntoAccount,
                transformationKind,
                state,
                tempState
            ));
    }

    public clipPath(instruction: Instruction): void{
        this.clipCurrentPath(instruction);
    }

    private incorporateDrawingInstruction(
        instruction: DrawingInstruction
    ): void{
        const area = instruction.getDrawnArea();
        if(area === empty){
            return;
        }
        const modifiedInstruction = instruction.getModifiedInstruction();
        this.addToPreviousInstructions(modifiedInstruction, area, instruction.build);
        if(!this.currentInstructionsWithPath){
            this.state = this.previousInstructionsWithPath.state;
        }
        this.onChange();
    }

    private addToPreviousInstructions(
        instruction: Instruction,
        area: Area,
        build: (instruction: Instruction) => ExecutableStateChangingInstructionSet
    ): void{
        const newInstruction = new InstructionsWithPositiveDrawnArea(build(instruction), area);
        newInstruction.setInitialStateWithClippedPaths(this.previousInstructionsWithPath.state);
        this.previousInstructionsWithPath.add(newInstruction);
    }

    private clipCurrentPath(instruction: Instruction): void{
        if(!this.currentInstructionsWithPath){
            return;
        }
        this.currentInstructionsWithPath.clipPath(instruction, this.state);
        this.state = this.currentInstructionsWithPath.state;
    }

    public addPathInstruction(pathInstruction: PathInstruction): void{
        if(this.currentInstructionsWithPath){
            this.currentInstructionsWithPath.addPathInstruction(pathInstruction, this.state);
        }
    }

    public closePath(): void{
        if(this.currentInstructionsWithPath){
            this.currentInstructionsWithPath.closePath();
        }
    }
    public moveTo(position: Position): void{
        if(this.currentInstructionsWithPath){
            this.currentInstructionsWithPath.moveTo(position, this.state);
        }
    }
    public canAddLineTo(position: Position): boolean{
        return !this.currentInstructionsWithPath || this.currentInstructionsWithPath.canAddLineTo(position, this.state);
    }
    public lineTo(position: Position): void{
        if(this.currentInstructionsWithPath){
            this.currentInstructionsWithPath.lineTo(position, this.state);
        }
    }
    public rect(rect: Rect): void{
        if(!this.currentInstructionsWithPath){
            return;
        }
        rect.addSubpaths(this.currentInstructionsWithPath, this.state)
    }
    public roundRect(
        rect: Rect,
		radii: number | DOMPointInit | Iterable<number | DOMPointInit> | undefined
    ): void{
        if(!this.currentInstructionsWithPath){
            return;
        }
        roundRect(this.currentInstructionsWithPath, rect, radii, this.state);
    }
    private intersects(area: Area): boolean{
        return this.previousInstructionsWithPath.intersects(area);
    }

    public clearContentsInsideArea(area: Area): void{
        this.previousInstructionsWithPath.clearContentsInsideArea(area);
        if(this.currentInstructionsWithPath){
            this.currentInstructionsWithPath.setInitialStateWithClippedPaths(this.previousInstructionsWithPath.state);
        }
    }

    public clearArea(x: number, y: number, width: number, height: number): void{
        const rectStrategy = getRectStrategy(x, y, width, height)
		const rectangle: Area = rectStrategy.getArea();
        if(!rectangle){
            return;
        }
        const transformedRectangle: Area = rectangle.transform(this.state.current.transformation)
        if(!this.intersects(transformedRectangle)){
            return;
        }
        this.clearContentsInsideArea(transformedRectangle);
        if(this.previousInstructionsWithPath.hasDrawingAcrossBorderOf(transformedRectangle)){
            this.previousInstructionsWithPath.addClearRect(rectangle, this.state, x, y, width, height);
            if(this.currentInstructionsWithPath){
                this.currentInstructionsWithPath.setInitialStateWithClippedPaths(this.state);
            }
        }
		this.onChange();
    }
    
    public execute(context: CanvasRenderingContext2D, rectangle: CanvasRectangle){
        if(this.previousInstructionsWithPath.length){
            this.previousInstructionsWithPath.execute(context, rectangle);
        }
        const latestVisibleState: InfiniteCanvasState = this.previousInstructionsWithPath.state;
        const stackLength = latestVisibleState.stack.length;
        for(let i = 0; i < stackLength; i++){
            context.restore();
        }
    }
}
