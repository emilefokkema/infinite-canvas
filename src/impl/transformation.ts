import { TransformationRepresentation } from "api/transformation-representation";
import { Point } from "./geometry/point"
import { PointAtInfinity } from "./geometry/point-at-infinity";

function stringifyNumber(value: number): string{
	return value.toFixed(10).replace(/\.?0+$/,'');
}

export class Transformation implements TransformationRepresentation{
	public scale: number;
	constructor(
		public a: number,
		public b: number,
		public c: number,
		public d: number,
		public e: number,
		public f: number
		){
			this.scale = Math.sqrt(a * d - b * c);
	}
	public getMaximumLineWidthScale(): number{
		const x1: number = this.a + this.c,
		y1: number = this.b + this.d,
		x2: number = this.a - this.c,
		y2: number = this.b - this.d;
		return Math.sqrt(Math.max(x1 * x1 + y1 * y1, x2 * x2 + y2 * y2));
	}
	public getRotationAngle(): number{
		const cos: number = this.a / this.scale;
		const sin: number = this.b / this.scale;
		if(cos === 0){
			if(sin > 0){
				return Math.PI / 2;
			}
			return 3 * Math.PI / 2;
		}
		const atan: number = Math.atan(sin / cos);
		if(cos > 0){
			if(sin > 0){
				return atan;
			}
			if(sin === 0){
				return 0;
			}
			return 2 * Math.PI + atan;
		}
		if(sin > 0){
			return Math.PI + atan;
		}
		if(sin === 0){
			return Math.PI;
		}
		return Math.PI + atan;
	}
	public applyToPointAtInfinity(pointAtInfinity: PointAtInfinity): PointAtInfinity{
		return {direction: this.untranslated().apply(pointAtInfinity.direction)};
	}
	public apply(point: Point): Point{
		return new Point(this.a * point.x + this.c * point.y + this.e, this.b * point.x + this.d * point.y + this.f);
	}
	public untranslated(): Transformation{
		const {x: originTransformedX, y: originTransformedY} = this.apply(Point.origin);
		return this.before(Transformation.translation(-originTransformedX, -originTransformedY));
	}
	public before(other: Transformation): Transformation{
		const a: number = other.a * this.a + other.c * this.b;
		const b: number = other.b * this.a + other.d * this.b;
		const c: number = other.a * this.c + other.c * this.d;
		const d: number = other.b * this.c + other.d * this.d;
		const e: number = other.a * this.e + other.c * this.f + other.e;
		const f: number = other.b * this.e + other.d * this.f + other.f;
		return new Transformation(a, b, c, d, e, f);
	}
	public equals(other: Transformation): boolean{
		return this.a === other.a &&
			   this.b === other.b &&
			   this.c === other.c &&
			   this.d === other.d &&
			   this.e === other.e &&
			   this.f === other.f;
	}
	public inverse(): Transformation{
		var det = this.a * this.d - this.b * this.c;
		if(det == 0){
			throw new Error("error calculating inverse: zero determinant");
		}
		const a: number = this.d / det,
			b: number = - this.b / det,
			c: number = - this.c / det,
			d: number = this.a / det,
			e: number = (this.c * this.f - this.d * this.e) / det,
			f: number = (this.b * this.e - this.a * this.f) / det;
		return new Transformation(a,b,c,d,e,f);
	}
	public static translation(dx: number, dy: number): Transformation{
		return new Transformation(1, 0, 0, 1, dx, dy);
	}
	public static scale(scale: number): Transformation{
		return new Transformation(scale, 0, 0, scale, 0, 0);
	}
	public static identity: Transformation = new Transformation(1,0,0,1,0,0);
	public static zoom(centerX: number, centerY: number, scale: number, translateX?: number, translateY?: number): Transformation{
		const oneMinScale: number = 1 - scale;
		if(translateX !== undefined){
			return new Transformation(scale, 0, 0, scale, centerX * oneMinScale + translateX, centerY * oneMinScale + translateY);
		}
		return new Transformation(scale, 0, 0, scale, centerX * oneMinScale, centerY * oneMinScale);
	}
	public static translateZoom(
		s1x: number,
		s1y: number,
		s2x: number,
		s2y: number,
		d1x: number,
		d1y: number,
		d2x: number,
		d2y: number): Transformation{
			const sdx: number = s2x - s1x;
			const sdy: number = s2y - s1y;
			const srsq = sdx * sdx + sdy * sdy;
			if(srsq === 0){
				throw new Error("divide by 0");
			}
			const ddx: number = d2x - d1x;
			const ddy: number = d2y - d1y;
			const drsq: number = ddx * ddx + ddy * ddy;
			const scale: number = Math.sqrt(drsq / srsq);
			return Transformation.zoom(s1x, s1y, scale, d1x - s1x, d1y - s1y);
	}
	public static rotation(centerX: number, centerY: number, radians: number): Transformation{
		const cos: number = Math.cos(radians);
		const sin: number = Math.sin(radians);
		const oneMinCos: number = 1 - cos;
		return new Transformation(
			cos,
			sin,
			-sin,
			cos,
			centerX * oneMinCos + centerY * sin,
			- centerX * sin + centerY * oneMinCos);
	}
	public static translateRotateZoom(
		s1x: number,
		s1y: number,
		s2x: number,
		s2y: number,
		d1x: number,
		d1y: number,
		d2x: number,
		d2y: number): Transformation{
			const sdx: number = s2x - s1x;
			const sdy: number = s2y - s1y;
			const det: number = sdx * sdx + sdy * sdy;
			if(det === 0){
				throw new Error("divide by 0");
			}
			const ddx: number = d2x - d1x;
			const ddy: number = d2y - d1y;
			const g: number = s1x * s2y - s1y * s2x;
			const h: number = s2x * sdx + s2y * sdy;
			const i: number = s1x * sdx + s1y * sdy;
			const a: number = (sdx * ddx + sdy * ddy) / det;
			const b: number = (sdx * ddy - sdy * ddx) / det;
			const c: number = -b;
			const d: number = a;
			const e: number = (d1x * h - d2x * i - g * ddy) / det;
			const f: number = (d1y * h - d2y * i + g * ddx) / det;
			return new Transformation(a, b, c, d, e, f);
	}
	public static create(representation: TransformationRepresentation): Transformation{
		if(representation instanceof Transformation){
			return representation;
		}
		const {a, b, c, d, e, f} = representation;
		return new Transformation(a, b, c, d, e, f);
	}
	public toString(): string{
		return `x: (${stringifyNumber(this.a)}, ${stringifyNumber(this.b)}), y: (${stringifyNumber(this.c)}, ${stringifyNumber(this.d)}), d: (${stringifyNumber(this.e)}, ${stringifyNumber(this.f)})`
	}
}
