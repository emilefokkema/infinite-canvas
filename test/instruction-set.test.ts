import {describe, it, expect, beforeEach, jest } from '@jest/globals';
import { Transformation } from "../src/transformation";
import { logInstruction } from "./log-instruction";
import { fillStyle } from "../src/state/dimensions/fill-stroke-style";
import { InfiniteCanvasInstructionSet } from "../src/infinite-canvas-instruction-set";
import { Point } from "../src/geometry/point";
import { HTMLCanvasRectangle } from "../src/rectangle/html-canvas-rectangle";
import { MockCanvasMeasurementProvider } from "./mock-canvas-measurement-provider";

describe("an instruction set", () => {
    let instructionSet: InfiniteCanvasInstructionSet;
    let onChangeSpy: jest.Mock;

    beforeEach(() => {
        onChangeSpy = jest.fn();
        instructionSet = new InfiniteCanvasInstructionSet(onChangeSpy, new HTMLCanvasRectangle(new MockCanvasMeasurementProvider(200, 200), {}));
    });

    describe("that begins drawing a path", () => {

        beforeEach(() => {
            instructionSet.beginPath();
            instructionSet.moveTo(new Point(0, 0));
            instructionSet.lineTo(new Point(3, 0));
            instructionSet.lineTo(new Point(0, 3));
        });

        it("should not have called onchange", () => {
            expect(onChangeSpy).not.toHaveBeenCalled();
        });

        describe("and then fills it", () => {

            beforeEach(() => {
                instructionSet.fillPath((context: CanvasRenderingContext2D) => {context.fill();});
            });

            it("should have called onchange", () => {
                expect(onChangeSpy).toHaveBeenCalledTimes(1);
            });

            describe("and then clears part of the drawing", () => {

                beforeEach(() => {
                    onChangeSpy.mockClear();
                    instructionSet.clearArea(0, 0, 1, 1);
                });

                it("should have called onchange", () => {
                    expect(onChangeSpy).toHaveBeenCalledTimes(1);
                });

                it("should have recorded a clearRect", () => {
                    expect(logInstruction((context: CanvasRenderingContext2D, transformation: Transformation) => {
                        instructionSet.execute(context, transformation);
                    })).toMatchSnapshot();
                });
            });
        });
    });

    describe("that changes state and draws a rectangle", () => {

        beforeEach(() => {
            instructionSet.changeState(s => fillStyle.changeInstanceValue(s, "#f00"));
            instructionSet.fillRect(0, 0, 1, 1, (context: CanvasRenderingContext2D, transformation: Transformation) => {
                context.fill();
            });
        });

        it("should have called onchange", () => {
            expect(onChangeSpy).toHaveBeenCalledTimes(1);
        });

        describe("and then partially clears that rectangle", () => {

            beforeEach(() => {
                onChangeSpy.mockClear();
                instructionSet.clearArea(0.5, 0, 2, 2);
            });

            it("should end up with a rectangle followed by a clearRect", () => {
                expect(logInstruction((context: CanvasRenderingContext2D, transformation: Transformation) => {
                    instructionSet.execute(context, transformation);
                })).toMatchSnapshot();
            });

            it("should have called onchange", () => {
                expect(onChangeSpy).toHaveBeenCalledTimes(1);
            });
        });

        describe("and then clears a rectangle", () => {

            beforeEach(() => {
                onChangeSpy.mockClear();
                instructionSet.clearArea(-1, -1, 3, 3);
            });

            it("should no longer have recorded the first rectangle", () => {
                expect(logInstruction((context: CanvasRenderingContext2D, transformation: Transformation) => {
                    instructionSet.execute(context, transformation);
                })).toMatchSnapshot();
            });

            it("should have called onchange", () => {
                expect(onChangeSpy).toHaveBeenCalledTimes(1);
            });
        });

        describe("and changes state and draws another rectangle", () => {

            beforeEach(() => {
                instructionSet.changeState(s => fillStyle.changeInstanceValue(s, "#00f"));
                instructionSet.fillRect(2, 0, 1, 1, (context: CanvasRenderingContext2D, transformation: Transformation) => {
                    context.fill();
                });
            });

            it("should have recorded everything in the right order", () => {
                expect(logInstruction((context: CanvasRenderingContext2D, transformation: Transformation) => {
                    instructionSet.execute(context, transformation);
                })).toMatchSnapshot();
            });
        });
    });
});
