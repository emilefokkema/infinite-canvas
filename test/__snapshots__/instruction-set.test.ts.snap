// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`an instruction set that begins drawing a path and then fills it and then clears part of the drawing should have recorded a clearRect 1`] = `
[
  "context.beginPath()",
  "context.moveTo(0,0)",
  "context.lineTo(3,0)",
  "context.lineTo(0,3)",
  "context.lineWidth = 1",
  "context.fill()",
  "context.save()",
  "context.transform(1,0,0,1,0,0)",
  "context.clearRect(0,0,1,1)",
  "context.restore()",
]
`;

exports[`an instruction set that changes state and draws a rectangle and changes state and draws another rectangle should have recorded everything in the right order 1`] = `
[
  "context.fillStyle = "#f00"",
  "context.lineWidth = 1",
  "context.beginPath()",
  "context.moveTo(0,0)",
  "context.lineTo(1,0)",
  "context.lineTo(1,1)",
  "context.lineTo(0,1)",
  "context.lineTo(0,0)",
  "context.closePath()",
  "context.moveTo(0,0)",
  "context.fill()",
  "context.fillStyle = "#00f"",
  "context.beginPath()",
  "context.moveTo(2,0)",
  "context.lineTo(3,0)",
  "context.lineTo(3,1)",
  "context.lineTo(2,1)",
  "context.lineTo(2,0)",
  "context.closePath()",
  "context.moveTo(2,0)",
  "context.fill()",
]
`;

exports[`an instruction set that changes state and draws a rectangle and then clears a rectangle should no longer have recorded the first rectangle 1`] = `[]`;

exports[`an instruction set that changes state and draws a rectangle and then partially clears that rectangle should end up with a rectangle followed by a clearRect 1`] = `
[
  "context.fillStyle = "#f00"",
  "context.lineWidth = 1",
  "context.beginPath()",
  "context.moveTo(0,0)",
  "context.lineTo(1,0)",
  "context.lineTo(1,1)",
  "context.lineTo(0,1)",
  "context.lineTo(0,0)",
  "context.closePath()",
  "context.moveTo(0,0)",
  "context.fill()",
  "context.save()",
  "context.transform(1,0,0,1,0,0)",
  "context.clearRect(0.5,0,2,2)",
  "context.restore()",
]
`;
